### PR TITLE
Shader and typings restructuring

### PR DESCRIPTION
**Pull request by birkskyum**
Previously we had a script called post-ts-build.js which would convert glsl shaders to .js and also make a .d.ts typings file for the whole library on every build.

I split up these into two files: 
- generate-shaders.ts
- generate-typings.ts

The generate-shaders.ts script will then populate ./src/shaders with generated .g.ts files that tsc and rollup will pick up like everything else.

### CODE REVIEW

**Comment by HarelM** [comment]
Shader files are used also in maplibre native, I'm not sure they can be moved without a proper notification to the guys maintaining it on slack.
@petr-pokorny-1 
Also adding them to source control is not great.
I would prefer to mive all the generated code to a specific library that is not source controlled and maybe generate them as a pist npm install script...
Just my 2 cents I guess...

**Comment by birkskyum** [comment]
I added the postinstall with generate-shaders, and remove the generated shaders. I can make a pr to maplibre-gl-native with a bump of this library and a change to the path. I noticed that the shaders are actually under version control in that repo: https://github.com/maplibre/maplibre-gl-native/tree/master/src/mbgl/programs

File: .gitignore [hunk 1]
```diff
@@ -19,6 +19,7 @@ test/integration/**/diff.png
 test/integration/**/test-results.xml
 test/integration/dist/**/*.js
 test/integration/dist/**/*.json
+src/shaders/*.glsl.ts
```
Comment (HarelM) [inline .gitignore:22]: I would prefer to start a folder called src/generated/*
or use something like file.g.ts to signal that this file is generated.

Comment (birkskyum) [reply .gitignore]: Good idea. I've change the name to .g.ts, as and updated .gitignore

File: package.json [hunk 2]
```diff
@@ -156,14 +155,13 @@
     "./rollup/build/tsc/src/util/web_worker.js": "./build/web_worker_replacement.js"
   },
   "scripts": {
-    "build-post-ts": "node build/post-ts-build.js ./rollup/build/tsc/src/shaders",
-    "build-post-ts-min": "node build/post-ts-build.js ./rollup/build/tsc/src/shaders true",
-    "build-tsc": "tsc --outDir rollup/build/tsc && npm run build-post-ts",
-    "build-tsc-min": "tsc --outDir rollup/build/tsc && npm run build-post-ts-min",
+    "generate-shaders": "node --loader ts-node/esm --experimental-specifier-resolution=node build/generate-shaders.js",
+    "generate-typings": "node --loader ts-node/esm --experimental-specifier-resolution=node build/generate-typings.js",
```
Comment (HarelM) [inline package.json:159]: Why do we need ts-node here?

File: build/generate-shaders.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import fs from 'fs';
+import glob from 'glob';
+
+let outputBaseDir = './src/shaders';
+let minify = true;
```
Comment (HarelM) [inline build/generate-shaders.ts:5]: If this is hardcoded true why keep it?

File: package.json (hunk 2 above)
Comment (HarelM) [reply package.json:159]: Also, if this only runs when releasing it might fail only then. I would prefer that this will run as part of the CI as well.

Comment (birkskyum) [reply package.json:159]: The ts-node/esm allows us to run typescript files directly. It converts it to .js on the go.

I added the generate-typings to the build workflow

File: build/generate-shaders.ts (hunk 3 above)
Comment (birkskyum) [reply build/generate-shaders.ts]: That is true, I have removed it. It takes virtually no extra time to minify, so I didn't see any reason not to.

Comment (xabbu42) [reply build/generate-shaders.ts]: I don't know anything about the specifics here, but wouldn't it be useful to disable minify for debuging?

File: package.json (hunk 2 above)
Comment (HarelM) [reply package.json:159]: The files that this commands run are `generate-shaders.js` and `generate-typings.js`.
It should be `generate-shaders.ts` then?
I'm a bit confused...

File: build/generate-shaders.ts (hunk 3 above)
Comment (HarelM) [reply build/generate-shaders.ts]: I'm not familiar with web gl code, but I'm not sure you can actually debug this code.
The minification is mainly white spaces here, so I'm not sure this really affects debugging if it's possible.

File: package.json (hunk 2 above)
Comment (birkskyum) [reply package.json]: it's more explicit now with .ts files

File: build/generate-shaders.ts (hunk 3 above)
Comment (birkskyum) [reply build/generate-shaders.ts]: Yes, the names aren't mangled here, so I figured it was okay even for debugging.

Comment (xabbu42) [reply build/generate-shaders.ts]: ah ok, thanks for the explanations

**Comment by birkskyum** [comment]
I believe all the mentioned points (thanks!) has been addressed, but let me know if there are still concerns holding this back or if I should do anything different.

**Comment by birkskyum** [comment]
Definitely, I will make a PR as soon at this is in.

**Comment by HarelM** [comment]
I world prefer to get an initial consent in slack first before merging this...

**Comment by HarelM** [comment]
Also I think the following file will no longer be needed after the PR, right?
https://github.com/maplibre/maplibre-gl-js/blob/main/src/types/glsl.d.ts

**Comment by birkskyum** [comment]
> Also I think the following file will no longer be needed after the PR, right? https://github.com/maplibre/maplibre-gl-js/blob/main/src/types/glsl.d.ts

I think we can take it out.

**Review by HarelM — APPROVED** [review]
Great work! Thanks!

**Pull request merged**